localstack/services/edge.py

```diff
@@ -437,46 +447,52 @@ def start_proxy(port, asynchronous=False):
     return proxy
 
 
-def start_edge(port=None, use_ssl=True, asynchronous=False):
-    if not port:
-        port = config.EDGE_PORT
-    if config.EDGE_PORT_HTTP and config.EDGE_PORT_HTTP != port:
-        do_start_edge(
-            config.EDGE_BIND_HOST,
-            config.EDGE_PORT_HTTP,
-            use_ssl=use_ssl,
-            asynchronous=True,
-        )
-    if port > 1024 or is_root():
-        return do_start_edge(config.EDGE_BIND_HOST, port, use_ssl, asynchronous=asynchronous)
-
-    # process requires privileged port but we're not root -> try running as sudo
-
-    class Terminator:
-        def stop(self, quiet=True):
-            try:
-                url = "http%s://%s:%s" % ("s" if use_ssl else "", LOCALHOST, port)
-                requests.verify_ssl = False
-                requests.post(url, headers={HEADER_KILL_SIGNAL: "kill"})
-            except Exception:
-                pass
-
-    # register a signal handler to terminate the sudo process later on
-    TMP_THREADS.append(Terminator())
-
-    # start the TCP proxy
-    env_vars = {
-        "DEBUG": os.environ.get("DEBUG", ""),
-        "EDGE_FORWARD_URL": config.get_edge_url(),
-        "EDGE_BIND_HOST": config.EDGE_BIND_HOST,
-    }
-    proxy_module = "localstack.services.edge"
-    proxy_args = ["proxy", str(port)]
-    return run_module_as_sudo(
-        module=proxy_module, arguments=proxy_args, env_vars=env_vars, asynchronous=asynchronous
-    )
+def split_list_by(lst: Iterable[T], predicate: Callable[[T], bool]) -> Tuple[List[T], List[T]]:
+    truthy, falsy = [], []
+
+    for item in lst:
+        if predicate(item):
+            truthy.append(item)
+        else:
+            falsy.append(item)
+
+    return truthy, falsy
 
 
+def start_edge(listen_str: str, use_ssl: bool = True, asynchronous: bool = False):
+    if listen_str:
+        listen = parse_gateway_listen(listen_str)
+    else:
+        listen = config.GATEWAY_LISTEN
+
+    if len(listen) == 0:
+        raise ValueError("no listen addresses provided")
+
+    # separate privileged and unprivileged addresses
+    unprivileged, privileged = split_list_by(listen, lambda addr: addr.is_unprivileged() or False)
+
+    # bind the gateway server to unprivileged addresses
+    edge_thread = do_start_edge(unprivileged, use_ssl=use_ssl, asynchronous=True)
+
+    # start TCP proxies for the remaining addresses
+    for address in privileged:
+        if is_root():
+            # just start the proxy
+            do_start_tcp_proxy(address, asynchronous=asynchronous)
+        else:
+            # escalate to root
+            run_module_as_sudo(
+                module="localstack.services.edge",
+                arguments=["proxy", "--gateway-listen", str(address)],
+                asynchronous=True,
+            )
+
+    if edge_thread is not None:
+        edge_thread.join()
+
+
+# TODO: rename to better represent what is going on, e.g. run_module_as_root
+# but this will break `ext`
```

Comment:
nit: Looking at this TODO: since there is a companion branch, you could just do this?

Reply:
I've removed this TODO for now. I don't feel strongly enough about it.